### ISSUE TITLE
Understanding Hinton’s Capsule Networks. Part III: Dynamic Routing Between Capsules.

### ISSUE THREAD

**Issue by sweaterr** [sweaterr/readinglist#342]

**Comment by sweaterr** [comment]
This is the third post in the series about a new type of neural network, based on capsules, called CapsNet.
CapsNet이라는 캡슐을 기반으로 한 새로운 유형의 신경망에 대한 시리즈 중 세 번째 게시물입니다.

I already talked about the intuition behind it, as well as what is a capsule and how it works.
나는 캡슐이 무엇인지, 캡슐이 어떻게 작동하는지뿐만 아니라 그 뒤에있는 직관에 대해서 이미 이야기했다.

In this post, I will talk about the novel dynamic routing algorithm that allows to train capsule networks.
이 글에서는 캡슐 네트워크를 훈련시키는 새로운 동적 라우팅 알고리즘에 대해 이야기 할 것입니다.

One of the earlier figures explaining capsules and routing between them.
캡슐을 설명하고 캡슐 사이를 라우팅하는 초기 수치 중 하나입니다.

Source.
출처.

As I showed in Part II, a capsule i in a lower-level layer needs to decide how to send its output vector to higher-level capsules j.
Part II에서 보았 듯이 하위 레벨의 캡슐 i는 출력 벡터를 더 높은 레벨의 캡슐 j로 보내는 방법을 결정해야합니다.

It makes this decision by changing scalar weight c_ij that will multiply its output vector and then be treated as input to a higher-level capsule.
출력 벡터를 곱한 다음 더 높은 레벨의 캡슐에 대한 입력으로 처리 할 스칼라 웨이트 c_ij를 변경하여이 결정을 내립니다.

Notation-wise, c_ij represents the weight that multiplies output vector from lower-level capsule i and goes as input to a higher level capsule j.
표기법에 따르면, c_ij는 하위 레벨 캡슐 i의 출력 벡터에 곱해지고 상위 레벨 캡슐 j에 대한 입력 값이되는 가중치를 나타냅니다.

Things to know about weights c_ij:  Each weight is a non-negative scalar For each lower level capsule i, the sum of all weights c_ij equals to 1 For each lower level capsule i, the number of weights equals to the number of higher-level capsules These weights are determined by the iterative dynamic routing algorithm The first two facts allow us to interpret weights in probabilistic terms.
가중치에 대해 알아야 할 사항 c_ij : 각 가중치는 음수가 아닌 스칼라입니다. 각 하위 캡슐 i에 대해 모든 가중치의 합 c_ij는 1입니다. 하위 캡슐 i마다 가중치 수는 상위 수준 수와 같습니다 캡슐 이러한 가중치는 반복적 인 동적 라우팅 알고리즘에 의해 결정됩니다. 처음 두 가지 사실을 통해 우리는 가중치를 확률 적으로 해석 할 수 있습니다.

Recall that the length a capsule’s output vector is interpreted as probability of existence of the feature that this capsule has been trained to detect.
캡슐의 출력 벡터가이 캡슐이 탐지하도록 훈련 된 특징의 존재 확률로 해석되는 길이를 상기하십시오.

Orientation of the output vector is the parametrized state of the feature.
출력 벡터의 방향은 피쳐의 매개 변수화 된 상태입니다.

So, in a sense, for each lower level capsule i, its weights c_ij define a probability distribution of its output belonging to each higher level capsule j.
그래서 어떤 의미에서 각 하위 캡슐 i에 대해 그 가중치 c_ij는 상위 캡슐 j에 속한 출력의 확률 분포를 정의합니다.

**Comment by sweaterr** [comment]
In the second line you will notice that there is a new coefficient b_ij that we haven’t seen before.
두 번째 줄에는 이전에 보지 못했던 새로운 계수 b_ij가 있음을 알 수 있습니다.

This coefficient is simply a temporary value that will be iteratively updated and, after the procedure is over, its value will be stored in c_ij.
이 계수는 반복적으로 업데이트 될 임시 값이며, 프로시저가 끝나면 그 값은 c_ij에 저장됩니다.

At start of training the value of b_ij is initialized at zero.
훈련 시작시 b_ij의 값은 0으로 초기화됩니다.

**Comment by sweaterr** [comment]
Line 3 says that the steps in 4–7 will be repeated r times (the number of routing iterations).
3 번 줄은 4-7 단계가 r 번 (반복 반복 수) 반복됩니다.

**Comment by sweaterr** [comment]
Step in line 4 calculates the value of vector c_i which is all routing weights for a lower level capsule i.
라인 4의 단계는 하위 레벨 캡슐 i에 대한 모든 라우팅 가중치 인 벡터 c_i의 값을 계산합니다.

This is done for all lower level capsules.
이것은 모든 하위 캡슐에 대해 수행됩니다.

Why softmax? Softmax will make sure that each weight c_ij is a non-negative number and their sum equals to one.
왜 softmax인가? Softmax는 각 무게 c_ij가 음수가 아니며 그 합이 1과 같음을 확인합니다.

Essentially, softmax enforces probabilistic nature of coefficients c_ij that I described above.
본질적으로, softmax는 위에서 설명한 계수 c_ij의 확률 적 특성을 강요합니다.

**Comment by sweaterr** [comment]
At the first iteration, the value of all coefficients c_ij will be equal, because on line two all b_ij are set to zero.
첫 번째 반복에서 모든 계수 c_ij의 값은 동일 할 것입니다. 왜냐하면 두 번째 줄에서 모든 b_ij가 0으로 설정되기 때문입니다.

For example, if we have 3 lower level capsules and 2 higher level capsules, then all c_ij will be equal to 0.5.
예를 들어, 우리가 3 개의 저수준 캡슐과 2 개의 상위 캡슐을 가지고 있다면, 모든 c_ij는 0.5가 될 것입니다.

The state of all c_ij being equal at initialization of the algorithm represents the state of maximum confusion and uncertainty: lower level capsules have no idea which higher level capsules will best fit their output.
알고리즘 초기화시 모든 c_ij의 상태는 최대 혼동 상태와 불확실성을 나타냅니다. 낮은 레벨 캡슐은 상위 캡슐이 출력에 가장 잘 맞는지 알지 못합니다.

Of course, as the process is repeated these uniform distributions will change.
물론 프로세스가 반복됨에 따라 이러한 균일 한 분포가 변경 될 것입니다.

**Comment by sweaterr** [comment]
After all weights c_ij were calculated for all lower level capsules, we can move on to line 5, where we look at higher level capsules.
모든 가중치 c_ij가 모든 하위 레벨 캡슐에 대해 계산 된 후 5 단계로 넘어 가서 상위 캡슐을 살펴볼 수 있습니다.

This step calculates a linear combination of input vectors, weighted by routing coefficients c_ij, determined in the previous step.
이 단계는 이전 단계에서 결정된 라우팅 계수 c_ij에 의해 가중 된 입력 벡터의 선형 조합을 계산합니다.

Intuitively, this means scaling down input vectors and adding them together, which produces output vector s_j.
직관적으로 이것은 입력 벡터의 크기를 줄인 다음 출력 벡터 s_j를 더하는 것을 의미합니다.

This is done for all higher level capsules.
이것은 모든 상위 캡슐에 대해 수행됩니다.

**Comment by sweaterr** [comment]
Next, in line 6 vectors from last step are passed through the squash nonlinearity, that makes sure the direction of the vector is preserved, but its length is enforced to be no more than 1.
다음 줄 6에서 마지막 단계의 벡터는 스쿼시 비선형 성을 통과하여 벡터의 방향이 유지되지만 길이는 1 이하로 유지됩니다.

This step produces the output vector v_j for all higher level capsules.
이 단계는 모든 상위 캡슐에 대한 출력 벡터 v_j를 생성합니다.

**Comment by sweaterr** [comment]
To summarize what we have so far: steps 4–6 simply calculate the output of higher level capsules.
지금까지 우리가 가지고있는 것을 요약하면, 단계 4-6은 단순히 상위 캡슐의 산출량을 계산합니다.

Step on line 7 is where the weight update happens.
7 번 단계는 체중 업데이트가 이루어지는 곳입니다.

This step captures the essence of the routing algorithm.
이 단계는 라우팅 알고리즘의 본질을 포착합니다.

This steps looks at each higher level capsule j and then examines each input and updates the corresponding weight b_ij according to the formula.
이 단계는 각 상위 캡슐 j를 살펴본 다음 각 입력을 검사하고 수식에 따라 해당 가중치 b_ij를 업데이트합니다.

The formula says that the new weight value equals to the old value plus the dot product of current output of capsule j and the input to this capsule from a lower level capsule i.
공식은 새로운 무게 값이 이전 값에 캡슐 j의 현재 출력과 더 낮은 레벨의 캡슐 i에서이 캡슐에 대한 입력의 내적을 더한 것과 같다고 말합니다.

The dot product looks at similarity between input to the capsule and output from the capsule.
내적은 캡슐에 대한 입력과 캡슐에서의 출력 사이의 유사성을 관찰합니다.

Also, remember from above, the lower level capsule will sent its output to the higher level capsule whose output is similar.
또한 위의 사항을 기억해두면 하위 레벨 캡슐은 출력이 비슷한 상위 캡슐로 출력을 보냅니다.

This similarity is captured by the dot product.
이 유사점은 내적에 의해 포착됩니다.

After this step, the algorithm starts over from step 3 and repeats the process r times.
이 단계 후에 알고리즘은 3 단계에서 시작하여 프로세스를 r 번 반복합니다.

**Comment by sweaterr** [comment]
After r times, all outputs for higher level capsules were calculated and routing weights have been established.
r 시간이 지난 후,보다 높은 수준의 캡슐에 대한 모든 산출물이 계산되었고, 운송 중량이 설정되었습니다.

The forward pass can continue to the next level of network.
순방향 패스는 다음 레벨의 네트워크로 계속 진행될 수 있습니다.

**Comment by sweaterr** [comment]
## Intuitive Example of Weight Update Step

**Comment by sweaterr** [comment]
In the figure on the left, imagine that there are two higher level capsules, their output is represented by purple vectors v1 and v2 calculated as described in previous section.
왼쪽의 그림에서 두 개의 상위 캡슐이 있다고 상상해보십시오. 이전 섹션에서 설명한대로 산출 된 자주색 벡터 v1 및 v2로 출력됩니다.

The orange vector represents input from one of the lower level capsules and the black vectors represent all the remaining inputs from other lower level capsules.
주황색 벡터는 하위 수준 캡슐 중 하나의 입력을 나타내고 검정 벡터는 다른 하위 수준 캡슐의 나머지 입력을 모두 나타냅니다.

**Comment by sweaterr** [comment]
We see that in the left part the purple output v1 and the orange input u_hat point in the opposite directions.
왼쪽에서 자색 출력 v1과 주황색 입력 u_hat이 반대 방향으로 향하는 것을 볼 수 있습니다.

In other words, they are not similar.
즉, 그들은 비슷하지 않습니다.

This means their dot product will be a negative number and as result routing coefficient c_11 will decrease.
즉, 내적이 음수가되고 결과로 라우팅 계수 c_11이 감소합니다.

In the right part, the purple output v2 and the orange input v_hat point in the same direction.
오른쪽 부분에서 보라색 출력 v2와 오렌지색 입력 v_hat은 같은 방향을 가리 킵니다.

They are similar.
그들은 비슷하다.

Therefore, the routing coefficient c_12 will increase.
따라서 라우팅 계수 c_12가 증가합니다.

This procedure is repeated for all higher level capsules and for all inputs of each capsule.
이 절차는 모든 상위 캡슐과 각 캡슐의 모든 입력에 대해 반복됩니다.

The result of this is a set of routing coefficients that best matches outputs from lower level capsules with outputs of higher level capsules.
이 결과는 하위 레벨 캡슐의 출력과 상위 캡슐의 출력을 가장 잘 조화시키는 일련의 라우팅 계수입니다.

**Comment by sweaterr** [comment]
## How Many Routing Iterations to Use?

The paper examined a range of values for both MNIST and CIFAR data sets.
이 논문은 MNIST 및 CIFAR 데이터 세트 모두에 대해 다양한 범위의 값을 조사했습니다.

Author’s conclusion is two-fold:  More iterations tends to overfit the data It is recommended to use 3 routing iterations in practice.
작성자의 결론은 두 가지입니다. 반복이 많을수록 데이터가 너무 많습니다. 실제로는 3 번의 반복 반복을 사용하는 것이 좋습니다.

**Comment by sweaterr** [comment]
## Conclusion

In this article, I explained the dynamic routing algorithm by agreement that allows to train the CapsNet.
이 기사에서는 CapsNet을 교육 할 수있는 계약에 따라 동적 라우팅 알고리즘을 설명했습니다.

The most important idea is that similarity between input and output is measured as dot product between input and output of a capsule and then routing coefficient is updated correspondingly.
가장 중요한 아이디어는 입력과 출력 사이의 유사성이 캡슐의 입력과 출력 사이의 내적 (dot product)으로 측정되고 라우팅 계수가 그에 따라 업데이트된다는 것입니다.

Best practice is to use 3 routing iterations.
3 가지 라우팅 반복을 사용하는 것이 가장 좋습니다.

In the next post, I will walk you through CapsNet architecture, where we will put together all pieces of the puzzle that we learned so far.
다음 글에서는 CapsNet 아키텍처를 살펴보면서 지금까지 배운 퍼즐의 모든 부분을 정리해 보겠습니다.
